### PR TITLE
feat(ESSNTL-4056): Add lastSeen filter

### PR DESCRIPTION
**Pull request by adonispuente**
[ESSNTL-4056](https://issues.redhat.com/browse/ESSNTL-4056)
This PR implements a new date range filter in inventory. The mvp for the filter is as follows :

- User can use the "Last seen" filter to select a date range.
- If a start date is selected, "updated_start" should be added to the GET /hosts query.
- If an end date is selected, "updated_end" should be added to the GET /hosts query.
- If no value is selected for the filter, neither "updated_start" nor "updated_end" should be added to the GET /hosts query, and the host results should remain unchanged.

### CODE REVIEW

File: src/components/filters/useLastSeenFilter.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+export const lastSeenFilterState = { lastSeenFilter: [] };
+export const LAST_SEEN_FILTER = 'LAST_SEEN_FILTER';
+export const lastSeenFilterReducer = (_state, { type, payload }) => ({
+    ...type === LAST_SEEN_FILTER && {
+        lastSeenFilter: payload
+    }
+});
+
+export const uselastSeenFilter = ([state, dispatch] = [lastSeenFilterState]) => {
```
Comment (gkarat) [inline src/components/filters/useLastSeenFilter.js:12]: nit-pick:
```suggestion
export const useLastSeenFilter = ([state, dispatch] = [lastSeenFilterState]) => {
```

File: src/components/filters/useLastSeenFilter.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    let [lastSeenStateValue, setLastSeenValue] = useState({});
+    const lastSeenValue = dispatch ? state.lastSeenFilter : [lastSeenStateValue];
+    const setValue = dispatch ? (newValue) => dispatch({ type: LAST_SEEN_FILTER, payload: newValue }) : setLastSeenValue;
+
+    const filter = {
+        label: 'Last seen',
+        value: 'last_seen',
+        type: 'radio',
```
Comment (gkarat) [inline src/components/filters/useLastSeenFilter.js:20]: I believe this is not really important, but the mocks don't have radio buttons:

image

if it doesn't require big changes, I think we can align it more to the mock's look

File: src/components/InventoryTable/EntityTableToolbar.js [hunk 3]
```diff
@@ -341,10 +360,25 @@ const EntityTableToolbar = ({
             ...enabledFilters.registeredWith ? [registeredFilter] : [],
             ...enabledFilters.rhcdFilter ? [rhcdFilterConfig] : [],
             ...enabledFilters.updateMethodFilter ? [updateMethodConfig] : [],
+            ...enabledFilters.lastSeenFilter ? [lastSeenFilter] : [],
             ...showTags && enabledFilters.tags ? [tagsFilter] : []
         ] : [],
         ...filterConfig?.items || []
     ];
+    const [startDate, setStartDate] = useState();
+    const [endDate, setEndDate] = useState();
```
Comment (gkarat) [inline src/components/InventoryTable/EntityTableToolbar.js:369]: nit-pick: could these two state values be replaced with the values from `lastSeenFilterValue`? Reducing the state variables for the toolbar component could help avoid extra updates/re-renders.

**Review by gkarat — APPROVED** [review]
great work! the PR is really neat. I could also test it successfully on my machine. there are some minor comments, but overall LGTM.

one idea: weird ancient dates like 199 will make the whole app crash 😁 probably we could limit the range of available "FROM" dates for a user? attaching video

[Screencast from 03-06-2023 01:45:38 PM.webm](https://user-images.githubusercontent.com/31385370/223114743-67492b20-b4d0-416e-9682-c01ea567fea9.webm)

File: src/components/filters/useLastSeenFilter.js (hunk 2 above)
Comment (adonispuente) [reply src/components/filters/useLastSeenFilter.js:20]: The ticket hasnt been updated, but there was a conversation with Margot from UX that gives the green light for radio-buttons to align more with the other apps!

File: src/components/InventoryTable/EntityTableToolbar.js (hunk 3 above)
Comment (mkholjuraev) [reply src/components/InventoryTable/EntityTableToolbar.js:369]: I am wondering if we can move all of these date-related functions to the filter itself. I believe you can access the filter values without saving them in a state.

Comment (mkholjuraev) [reply src/components/InventoryTable/EntityTableToolbar.js:369]: If we keep these states in the toolbar component, we need to re-render the whole toolbar rather than the filter only

Comment (adonispuente) [reply src/components/InventoryTable/EntityTableToolbar.js:369]: > 
Hmm it seems the api breaks when you send a date like that, good catch. Going to ask the backend what it should be limited too. As for the other suggestions, they are all implemented locally and I will push it up all together once I hear back as to include all changes in one commit. Thank you both for taking a look!

Comment (gkarat) [reply src/components/InventoryTable/EntityTableToolbar.js]: We still can simply fix the issue on our end by limiting the acceptable time range.. and not passing abnormal values to the back end

Comment (adonispuente) [reply src/components/InventoryTable/EntityTableToolbar.js]: When we pass in an abnormal value like the 199 you had, the datePicker converts it to a readable value , like 0199, which is technically correct.
The backend just wasnt aware of this edge case possibility and doesnt know what the minimum value it will accept is. Once the come back to me with it I can update the PR with that validating edge case and/or the backend will update to accept such edge case.

**Comment by adonispuente** [comment]
ci test are failing for the same issue discussed in slack. All test pass locally, will merge today
